### PR TITLE
Add a new `SdpaFwdOp` IR node for Flash Attention

### PR DESCRIPTION
**Pull request by Priya2698**
Issue #2278.
~This PR adds the new node with same functionality as `torch.nn.functional.scaled_dot_product_attention`, and enables scheduling it through `ExprEvalScheduler`.~

Based on the PR discussions, this PR is repurposed to introduce a new IR node `SdpaFwdOp` for scaled dot product flash attention forward (see #2278 for details).
This PR does not include changes to the scheduler.

The next PRs will:
1. Add mapping for producer-consumer in root_domain_map and enable this op in `ExprEvalScheduler`.
2. Python API
3. Add a node for backward pass similar to this forward node.

After the completion of these tasks, we also aim at introducing Memory Efficient Attention.

### CODE REVIEW

File: csrc/ir/internal_nodes.h [hunk 1]
```diff
@@ -2327,4 +2327,77 @@ class LinearOp : public Expr {
...
+  using Expr::Expr;
+
+  SdpaOp(
+      IrBuilderPasskey,
+      Val* out,
+      Val* query,
+      Val* key,
+      Val* value,
+      Val* attn_mask,
+      double dropout_p,
```
Comment (jacobhinkle) [inline csrc/ir/internal_nodes.h:2343]: Should we instead let `dropout_p` be a `Val*`?

File: csrc/ir/internal_nodes.h [hunk 2]
```diff
@@ -2327,4 +2327,77 @@ class LinearOp : public Expr {
...
+      bool is_causal,
+      std::optional<double> scale);
```
Comment (jacobhinkle) [inline csrc/ir/internal_nodes.h:2345]: Same. I think this should be a scalar `Val*`, since it will commonly be a function of the input dimensions.

File: tests/cpp/utils.h [hunk 3]
```diff
@@ -683,4 +684,6 @@ at::Tensor atBiasEpilogue(const at::Tensor& tensor, const at::Tensor& bias);
 // Get the number of SMs on the current device
 int64_t getNumSMs();
 
+const bool checkMapped(const ValGraph& vg, IterDomain* x, IterDomain* y);
```
Comment (jjsjann123) [inline tests/cpp/utils.h:687]: nitpick: a const value doesn't do anything on a function's return

File: csrc/ir/internal_nodes.h (hunk 1 above)
Comment (jjsjann123) [reply csrc/ir/internal_nodes.h:2343]: I agree with @jacobhinkle here.

Since we already allow things like `attn_mask` to be dynamic, we might as well also extend that to all arguments. Including `is_causal`.

IIUC, given current implementation, if we are going to want to schedule and codegen sdpa in the future, we'll have to concretize on attn_mask anyway, since the computational pattern would change based on whether or not we'll have an attn_mask. Why stop there and not extend this all the way to other arguments?

File: csrc/ir/internal_nodes.h [hunk 4]
```diff
@@ -2327,4 +2327,77 @@ class LinearOp : public Expr {
   }
 };
 
+// SDPA node with same functionality as F.scaled_dot_product_attention
+// https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/internal_nodes.h:2331]: Thanks for raising this point. The current `SdpaOp` node will work only for inference.
For training: we would need to create two separate ops for forward and backward with different function signatures.

Another issue that @wujingyue pointed out is, that the current ATen evaluation does not look at whether the inputs are sharded which would need some modifications to the `evaluate` method. This is broader issue, since it applies to all other ops as well but particularly to `MatmulOp` and `LinearOp`, where we use the ATen fallback path primarily..

[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/ir/internal_nodes.h:2331]: Just want to point out that, for our ExprEvaluator, returning the intermediates (residuals?!) for backward isn't a big deal, but if we want to decompose and schedule it, we might want to be more careful about it.

Ideally speaking, nvfuser should be able to trim un-used outputs. So maybe we should switch, but which signature is the right one to target here? PyTorch itself has so many different variants and each has a different signature:
https://github.com/pytorch/pytorch/blob/a6b994ed5467d4df8320cbae51cba6a98ffb139c/aten/src/ATen/native/transformers/attention.cpp#L665-L706
https://github.com/pytorch/pytorch/blob/a6b994ed5467d4df8320cbae51cba6a98ffb139c/tools/autograd/derivatives.yaml#L2806-L2829

[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/internal_nodes.h:2331]: I summarized the current plan we settled on here: https://github.com/NVIDIA/Fuser/issues/2278

**Comment by Priya2698** [comment]
!build

File: csrc/ir/nodes.cpp [hunk 5]
```diff
@@ -4336,4 +4336,135 @@ std::vector<PolymorphicValue> LinearOp::evaluate(
...
+  indent(ss, indent_size) << attn_out()->toString() << "\n";
+  indent(ss, indent_size + 1) << " = sdpa(" << query()->toString() << ",\n";
+  indent(ss, indent_size + 1) << "          " << key()->toString() << ",\n";
+  indent(ss, indent_size + 1) << "          " << value()->toString() << ",\n";
+  indent(ss, indent_size + 1)
+      << "          dropout_p = " << dropout_p() << ",\n";
+  indent(ss, indent_size + 1) << "          is_causal = " << is_causal();
+  if (scale() != nullptr) {
+    indent(ss, indent_size + 1) << ",\n          scale = " << scale();
+  }
```
Comment (jacobhinkle) [inline csrc/ir/nodes.cpp:4390]:
```suggestion
  indent(ss, indent_size + 1)
      << "          dropout_p = " << dropout_p()->toInlineString() << ",\n";
  indent(ss, indent_size + 1) << "          is_causal = " << is_causal()->toInlineString();
  if (scale() != nullptr) {
    indent(ss, indent_size + 1) << ",\n          scale = " << scale()->toInlineString();
  }
```
It's only safe to use `toInlineString()` here if these vals cannot be `TensorView`. It seems like they should always be scalar to me. Can you confirm? If so, should we have some code in the ctor that checks that I couldn't for example pass a `TensorView` for `is_causal`?

File: csrc/ir/nodes.cpp [hunk 6]
```diff
@@ -4336,4 +4336,135 @@ std::vector<PolymorphicValue> LinearOp::evaluate(
...
+      [output,
+       log_sumexp,
+       cum_seq_q,
+       cum_seq_k,
+       query_seq_len,
+       key_seq_len,
+       philox_seed,
+       philox_offset,
+       debug_attn_mask] =
+          at::_scaled_dot_product_flash_attention(
```
Comment (jacobhinkle) [inline csrc/ir/nodes.cpp:4442]: Do we have a plan for handling the other execution modes? I'm assuming there's no higher level op in ATen and that that logic is in the python code for PyTorch?

**Review by jacobhinkle — COMMENTED** [review]
Is it correct to say that before accepting this in ExprEval scheduler, we need to handle the other execution modes?

**Comment by Priya2698** [comment]
> Is it correct to say that before accepting this in ExprEval scheduler, we need to handle the other execution modes?

No. I mainly separated them to handle the mapping, and the ID graphs workarounds separately than the node to reduce the scope of this PR.

As we discussed in today's meeting, at the moment we only plan on supporting Flash Attention to support multi-GPU development. Once we support Flash Attention, we can revisit, if we need to add Memory Efficient Attention as well. There could be a few ways:
1. Plumbing down the backend info from Thunder and using that within our nodes: While the two implementations have different function signatures, there are overlaps and hence, one possibility is to use a superset of the inputs and outputs. The other design here would be distinct nodes for each implementation.
2. We make the decision about the backend within nvFuser using the same logic as ATen/Thunder. See: https://github.com/Lightning-AI/lightning-thunder/blob/9f0c50cc6df187cf5fd2e31240690fe2b5e9ccc1/thunder/executors/sdpaex.py#L618-L680

File: csrc/ir/nodes.cpp (hunk 6 above)
Comment (Priya2698) [reply csrc/ir/nodes.cpp:4442]: What do you mean by higher level op? `at::scaled_dot_product_attention` has the backend logic within its implementation and switches between different modes. We cannot use this implementation since we need the intermediates from these underlying functions for the backward pass.

**Comment by jacobhinkle** [comment]
> > Is it correct to say that before accepting this in ExprEval scheduler, we need to handle the other execution modes?
> 
> No. I mainly separated them to handle the mapping, and the ID graphs workarounds separately than the node to reduce the scope of this PR.
> 
> As we discussed in today's meeting, at the moment we only plan on supporting Flash Attention to support multi-GPU development. Once we support Flash Attention, we can revisit, if we need to add Memory Efficient Attention as well. There could be a few ways:
> 1. Plumbing down the backend info from Thunder and using that within our nodes: While the two implementations have different function signatures, there are overlaps and hence, one possibility is to use a superset of the inputs and outputs. The other design here would be distinct nodes for each implementation.
> 2. We make the decision about the backend within nvFuser using the same logic as ATen/Thunder. See: https://github.com/Lightning-AI/lightning-thunder/blob/9f0c50cc6df187cf5fd2e31240690fe2b5e9ccc1/thunder/executors/sdpaex.py#L618-L680

I dont understand how to do partial support. If we are given inputs that we cannot evaluate with flashattention, what will happen?

**Comment by Priya2698** [comment]
> > > Is it correct to say that before accepting this in ExprEval scheduler, we need to handle the other execution modes?
> > 
> > 
> > No. I mainly separated them to handle the mapping, and the ID graphs workarounds separately than the node to reduce the scope of this PR.
> > As we discussed in today's meeting, at the moment we only plan on supporting Flash Attention to support multi-GPU development. Once we support Flash Attention, we can revisit, if we need to add Memory Efficient Attention as well. There could be a few ways:
> > 
> > 1. Plumbing down the backend info from Thunder and using that within our nodes: While the two implementations have different function signatures, there are overlaps and hence, one possibility is to use a superset of the inputs and outputs. The other design here would be distinct nodes for each implementation.
> > 2. We make the decision about the backend within nvFuser using the same logic as ATen/Thunder. See: https://github.com/Lightning-AI/lightning-thunder/blob/9f0c50cc6df187cf5fd2e31240690fe2b5e9ccc1/thunder/executors/sdpaex.py#L618-L680
> 
> I dont understand how to do partial support. If we are given inputs that we cannot evaluate with flashattention, what will happen?

We will only accept the op if the backend identified is Flash Attention in Thunder: https://github.com/Lightning-AI/lightning-thunder/blob/9f0c50cc6df187cf5fd2e31240690fe2b5e9ccc1/thunder/executors/sdpaex.py#L618-L680. 

Do you think this will not be sufficient?

File: csrc/ir/nodes.cpp (hunk 6 above)
Comment (jacobhinkle) [reply csrc/ir/nodes.cpp:4442]: > We cannot use this implementation since we need the intermediates from these underlying functions for the backward pass.

Ah ok. It looks like that logic is available in this function https://github.com/pytorch/pytorch/blob/c3949b20a14875bf22934246404eb3cb497fb9cc/aten/src/ATen/native/transformers/sdp_utils_cpp.cpp#L60
Could we use `sdp::select_sdp_backend_cpp` here in the future to dispatch to the other modes directly?

**Comment by jacobhinkle** [comment]
> Do you think this will not be sufficient?

Makes sense to me. The logic is taking place in thunder to dispatch to flash attention, which seems fine.

File: csrc/ir/nodes.cpp (hunk 6 above)
Comment (Priya2698) [reply csrc/ir/nodes.cpp:4442]: Yes, that is the second method we could use if we don't want to make the decision within Thunder.

**Comment by naoyam** [comment]
Is anyone still reviewing this PR? @jacobhinkle?

**Review by jacobhinkle — APPROVED** [review]
LGTM once this is addressed: https://github.com/NVIDIA/Fuser/pull/2294#discussion_r1626546770

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
Restarted the failed tests: https://gitlab-master.nvidia.com/dl/pytorch/fuser-gh-mirror/-/pipelines/15652567

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
The failing tests look unrelated.
